src/main/java/net/minecraftforge/common/capabilities/IAttachedCapabilityProvider.java

```diff
@@ -0,0 +1,137 @@
+     */
+    @NotNull Capability<C> getCapability(@Nullable Direction direction);
+
+    /**
+     * @see {@link ICapabilityProvider#invalidateCaps()} for documentation.
+     */
+    void invalidateCaps();
+
+    /**
+     * @see {@link ICapabilityProvider#reviveCaps()} for documentation.
+     */
+    void reviveCaps();
+
+    /**
+     * Serialize this provider to NBT, if applicable.<br>
+     * You should always serialize your state, even if {@link #invalidateCaps()} has been called.
+     * @return A {@link CompoundTag} representing the serialized form of this provider, or null, if not serializable.
+     */
+    default @Nullable CompoundTag serializeNBT()
+    {
+        return null;
+    }
+
+    /**
+     * Reads this provider from NBT.<br>
+     * This method will only be called if this provider was serialized via {@link #serializeNBT()}.<br>
+     * You should always deserialize your state, even if {@link #invalidateCaps()} has been called.
+     * @param tag The serialized tag from {@link #serializeNBT()}.
+     */
+    default void deserializeNBT(CompoundTag tag)
+    {
+    }
+
+    /**
+     * Specialized subclass of {@link IAttachedCapabilityProvider} for use with {@link ItemStack}.
+     * Has special functions that allow for the complexity of merging and copying stacks.
+     *
+     * @param <C> The type of the capability.
+     */
+    public static interface IComparableCapabilityProvider<C, O extends ICapabilityProvider> extends IAttachedCapabilityProvider<C, O>
+    {
+        /**
+         * <b>This method is performance-critical!  Try to make this as fast as possible while ensuring correctness.</b>
+         * <p>
+         * For {@link ItemStack}s to stack, it must be validated that they are equivalent.<br>
+         * This means validating that all attached capabilities are equivalent as well.<br>
+         * This method allows attached capabilities to define if they are equivalent or not.
+         * <p>
+         * Specifically, this method compares this provider to another instance of this provider
+         * on another object, or null (if it is not present on the other object).
+         * <p>
+         * If your capability must be equivalent for itemstacks to merge, you MUST override this method.
+         * <p>
+         * It can be assumed that all other parts of the itemstack match by the time this is called.  As such, if your
+         * cap does not store unserialized data (outside of stack NBT), then you can either return <code>true</code> or 
+         * <code>other != null</code>.
+         * 
+         * @param other Another instance of this provider from another object, or null, if it was not present.
+         * @return true if these providers are equivalent, otherwise false.
+         */
+        boolean isEquivalentTo(@Nullable IComparableCapabilityProvider<C, O> other);
+    }
+
+    /**
+     * Specialized subclass of {@link IAttachedCapabilityProvider} for use with {@link ItemStack}.
```

Comment:
This comment (and most of the ones on this interface) seem to be specific to `ItemStack`, yet in `AttachCapabilitiesEvent.Players`, it specifies that these capabilities must be copyable. I would think this should be reflected in the documentation here.